code/modules/telesci/dses.dm

```diff
@@ -0,0 +1,291 @@
+	icon_state = "dses"
+	icon = 'icons/obj/telescience.dmi'
+	var/obj/item/weapon/cell/C = null
+	var/module_limit = 1
+	var/list/module_list = list()
+	var/list/locations = list()
+	var/pulse_range = 25
+	var/pulse_cost = 250
+	var/pulse_between = 3 SECONDS
+	var/last_pulse
+	var/last_direction
+	var/last_distance
+	var/list/positive_locations = list()
+	var/auto_pulse = 0
+
+/obj/item/device/dses/attack_self(var/mob/user as mob)
+	if(C)
+		return menu_open(user)
+	else
+		to_chat(user, "<span class = 'notice'>The screen remains dark.</span>")
+		return
+
+/obj/item/device/dses/attackby(obj/item/weapon/W, mob/user)
+	if(istype(W, /obj/item/weapon/dses_module))
+		install_module(W, user)
+	if(istype(W, /obj/item/weapon/cell))
+		if(C)
+			to_chat(user, "<span class = 'notice'>There is already a cell installed.</span>")
+		else
+			if(user.drop_item(W, src))
+				to_chat(user, "<span class = 'notice'>You install \the [W] into the battery slot.</span>")
+				C = W
+	..()
+
+/obj/item/device/dses/proc/menu_open(var/mob/user as mob)
+	var/obj/item/device/dses/t = ""
+	if(C)
+		if(!module_in_list("utoPNG"))
+			t += "<BR><A href='?src=\ref[src];pulse=1'>Pulse</A> "
+		else
+			t += "<BR><A href='?src=\ref[src];tog_pulse=1'>Toggle Pulse:\[[auto_pulse]\]</A>"
+		t += "<BR>Current cell charge: [C.charge]/[C.maxcharge]"
+		t += "<BR>Current device statistics"
+		t += "<BR>Pulse cost: [pulse_cost]"
+		t += "<BR>Pulse range: [pulse_range]"
+		if(module_in_list("dirGET"))
+			t += "<BR><B>Last logged direction: [last_direction]</B>"
+		if(module_in_list("getDST"))
+			t += "<BR><B>Distance from previous ping: [last_distance]</B>"
+		if(module_in_list("crdLOG") && positive_locations.len)
+			t += "<BR>==Logged successful pings=="
+			var/current_loc_itt
+			for(var/atom/A in positive_locations)
+				current_loc_itt ++
+				t += "<BR>\the [A] in [A.loc] at [A.x-WORLD_X_OFFSET[A.z]] - [A.y-WORLD_Y_OFFSET[A.z]] - [A.z]"
+				t += " === <A href='?src=\ref[src];wipe=[current_loc_itt]'>Wipe log</A>"
+			t += "<BR><A href='?src=\ref[src];wipe=0'>Wipe All</A>"
+		t += "<BR><B>Modules installed</B>"
+		if(module_list.len)
+			var/current_module_num = 0
+			for(var/obj/item/weapon/dses_module/D in module_list)
+				current_module_num ++
+				t += "<BR>Module: [D.module_name]"
+				t += "<BR><A href='?src=\ref[src];eject=[current_module_num]'>Eject Module</A>"
+		else
+			t += "<BR>No modules detected"
+		t += "<BR>Module slots [module_list.len]/[module_limit]"
+		t += "<BR><A href='?src=\ref[src];eject_cell=1'>Eject Cell</A>"
+	else
+		t += "<B>No cell detected.</B>"
+	var/datum/browser/popup = new(user, "DSES", name, 300, 450)
+	popup.set_content(t)
+	popup.set_title_image(user.browse_rsc_icon(src.icon, src.icon_state))
+	popup.open()
+
+/obj/item/device/dses/Topic(href, href_list)
+	..()
+	usr.set_machine(src)
+	if(href_list["pulse"])
+		if (usr.get_active_hand() != src || usr.stat)
```

Comment:
This kind of sanity should not go into a single href. Make it in the entire proc.

Also: use `isConscious`